### PR TITLE
fix: incomplete blocks removal during delete operation

### PR DESCRIPTION
**Pull request by IkbelTalebHssan**
# Motivation

This PR includes a fix for blocks bulk removal in the visual editor.

Fixes [# (issue)](https://github.com/Hexastack/Hexabot/issues/266)

# Type of change:

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added unit tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: frontend/src/components/visual-editor/v2/Diagrams.tsx [hunk 1]
```diff
@@ -380,48 +380,52 @@ const Diagrams = () => {
       } else {
         // Block Delete Case
         const ids = id.includes(",") ? id.split(",") : [id];
-        const deletePromises = ids.map((id) => {
-          const block = getBlockFromCache(id);
 
-          return deleteBlock(id, {
+        for (const blockId of ids) {
```
Comment (marrouchi) [inline frontend/src/components/visual-editor/v2/Diagrams.tsx:384]: Thanks @IkbelTalebHssan 

Can we rather implement deleteMany() on the API side ?

**Review by marrouchi — CHANGES_REQUESTED** [review]

File: frontend/src/components/visual-editor/v2/Diagrams.tsx (hunk 1 above)
Comment (IkbelTalebHssan) [reply frontend/src/components/visual-editor/v2/Diagrams.tsx]: Thanks :pray: 
I made some updates can you please check again :pray:

**Review by marrouchi — APPROVED** [review]

File: frontend/src/components/visual-editor/v2/Diagrams.tsx [hunk 2]
```diff
@@ -299,135 +306,166 @@ const Diagrams = () => {
...
+      },
+    );
+  };
+  const handleBlockDeletion = async (blockIds: string[]) => {
+    if (blockIds.length > 1) {
+      await deleteMultipleBlocks(blockIds);
+    } else {
+      await deleteSingleBlock(blockIds[0]);
+    }
+  };
```
Comment (medchedli) [inline frontend/src/components/visual-editor/v2/Diagrams.tsx:370]: handleBlockDeletion could be simplified by directly calling deleteBlocks for both single and multiple block deletions.

File: api/src/chat/controllers/block.controller.ts [hunk 3]
```diff
@@ -293,4 +293,29 @@ export class BlockController extends BaseController<
...
+
+  /**
+   * Deletes multiple blocks by their IDs.
+   * @param ids - IDs of blocks to be deleted.
+   * @returns A Promise that resolves to the deletion result.
+   */
+  @CsrfCheck(true)
+  @Delete('')
+  @HttpCode(204)
+  async deleteMany(@Body('ids') ids: string[]): Promise<DeleteResult> {
```
Comment (medchedli) [inline api/src/chat/controllers/block.controller.ts:305]: Renaming the method from deleteMany to deleteBlocksByIds will be more suitable.

File: frontend/src/components/visual-editor/v2/Diagrams.tsx [hunk 4]
```diff
@@ -299,135 +306,166 @@ const Diagrams = () => {
...
+    if (!ids || ids?.length === 0) {
+      return;
+    }
+    const isLink = ids[0].length === 36;
 
-      deleteCallbackRef.current?.();
-      deleteCallbackRef.current = () => {};
-      deleteDialogCtl.closeDialog();
+    if (isLink) {
+      await handleLinkDeletion(ids[0]);
```
Comment (medchedli) [inline frontend/src/components/visual-editor/v2/Diagrams.tsx:463]: Currently, the implemented logic only covers deleting one Link, while the user could select more than one link at the same time or could select many blocks and many links at the same time and trigger the action to delete them.

Comment (abdou6666) [reply frontend/src/components/visual-editor/v2/Diagrams.tsx:463]: I discussed it with @marrouchi and we decided on keeping the same logic for now and in the future update it

Comment (marrouchi) [reply frontend/src/components/visual-editor/v2/Diagrams.tsx:463]: @medchedli We will handle this issue in a seperate PR, can you open an issue about blocks VS links deletion ?

File: api/src/chat/controllers/block.controller.ts (hunk 3 above)
Comment (marrouchi) [reply api/src/chat/controllers/block.controller.ts:305]: Let's keep the naming like this, cause all the names are standardized across all controllers.

**Comment by medchedli** [comment]
Added a new issue to document the problem with deleting multiple links #363

**Pull request merged**